### PR TITLE
CRM-20663: Support recording of payment against fully paid contribution

### PR DESCRIPTION
**Pull request by monishdeb**
* [CRM-20663: Support recording of payment against fully paid contribution](https://issues.civicrm.org/jira/browse/CRM-20663)

### CODE REVIEW

**Comment by monishdeb** [comment]
Before additional payment:
image

After additional payment of $20 against fully paid contribution:
image

'Record Payment' is now available besides Total Amount in Contribution View Page:
image

File: CRM/Contribute/BAO/Contribution.php [hunk 1]
```diff
@@ -3858,7 +3858,86 @@ public static function recordAdditionalPayment($contributionId, $trxnsData, $pay
     $params['skipLineItem'] = TRUE;
     $trxnsData['trxn_date'] = !empty($trxnsData['trxn_date']) ? $trxnsData['trxn_date'] : date('YmdHis');
     $arAccountId = CRM_Contribute_PseudoConstant::getRelationalFinancialAccount($contributionDAO->financial_type_id, 'Accounts Receivable Account is');
-    if ($paymentType == 'owed') {
+    $fiancialItemPaidStatusID = CRM_Core_PseudoConstant::getKey('CRM_Financial_DAO_FinancialItem', 'status_id', 'Paid');
```
Comment (JoeMurray) [inline CRM/Contribute/BAO/Contribution.php]: please correct typo $financialItemPaidStatusID

File: CRM/Contribute/BAO/Contribution.php [hunk 2]
```diff
@@ -3858,7 +3858,86 @@ public static function recordAdditionalPayment($contributionId, $trxnsData, $pay
...
+    if ($paymentType == 'overpaid') {
```
Comment (JoeMurray) [inline CRM/Contribute/BAO/Contribution.php]: $paymentType as variable name for 'owed' or 'refund' makes some sense. 'overpaid' seems not quite right. Let's change to 'owed', 'more than owed', and 'refund'.

File: CRM/Contribute/BAO/Contribution.php [hunk 3]
```diff
@@ -3858,7 +3858,86 @@ public static function recordAdditionalPayment($contributionId, $trxnsData, $pay
...
+
+      // add financial trxn for the overpaid amount
+      $financialTrxn = CRM_Contribute_BAO_Contribution::recordFinancialAccounts($params, $trxnsData);
+
+      $lineItems = CRM_Price_BAO_LineItem::getLineItemsByContributionID($contributionDAO->id);
+      $lineItemID = NULL;
+      // if line item found then lookout for the one which is configured  with contribution
+      // quick config price field, whose line_total will be updated with overpaid amount.
+      if (!empty($lineItems)) {
+        // find line item params which is used later to store the over paid amount
```
Comment (JoeMurray) [inline CRM/Contribute/BAO/Contribution.php]: NO. The overpaid amount is not to be allocated to line items. Line items can only be paid up to the obligation they create. The overpaid amount is simply not allocated to line items.

File: CRM/Contribute/BAO/Contribution.php [hunk 4]
```diff
@@ -3858,7 +3858,86 @@ public static function recordAdditionalPayment($contributionId, $trxnsData, $pay
...
+        $priceSetId = CRM_Core_DAO::getFieldValue('CRM_Price_DAO_PriceSet', 'default_contribution_amount', 'id', 'name');
+        foreach ($lineItems as $id => $lineItem) {
+          if ($lineItem['price_set_id'] == $priceSetId) {
+            $lineItemID = $id;
+            $lineItemParams = array(
+              'id' => $id,
+               // update both line_total and unit_price with additional overpaid amount
```
Comment (JoeMurray) [inline CRM/Contribute/BAO/Contribution.php]: Please correct to exclude overpayment amount

File: CRM/Contribute/BAO/Contribution.php [hunk 5]
```diff
@@ -3858,7 +3858,86 @@ public static function recordAdditionalPayment($contributionId, $trxnsData, $pay
...
+              'line_total' => $lineItem['line_total'] + $trxnsData['total_amount'],
+              'unit_price' => $lineItem['unit_price'] + $trxnsData['total_amount'],
+            );
+            civicrm_api3('LineItem', 'create', $lineItemParams);
+            break;
+          }
+        }
+      }
+      // If lineitem not found then create a new line item referring to default
```
Comment (JoeMurray) [inline CRM/Contribute/BAO/Contribution.php]: I don't understand this. How can there ever be a contribution without a line item? If there is, it is an error condition, and we shouldn't try to correct by inserting a line item.

File: CRM/Contribute/BAO/Contribution.php [hunk 6]
```diff
@@ -3858,7 +3858,86 @@ public static function recordAdditionalPayment($contributionId, $trxnsData, $pay
...
+            'price_field_value_id' => $values['priceFieldValueID'],
+            'label' => $values['label'],
+          );
+          // add new line item entry and fetch its ID, later used for adding new financial item
+          $lineItemID = CRM_Utils_Array::value('id', civicrm_api3('LineItem', 'create', $lineItemParams));
+          break;
+        }
+      }
+
+      // create financial item for the overpaid amount linking to the
```
Comment (JoeMurray) [inline CRM/Contribute/BAO/Contribution.php]: please remove

**Comment by eileenmcnaughton** [comment]
@monishdeb @JoeMurray I think we need to clarify (& agree) the spec here. I agree with the logic that it should be possible to add a payment to a contribution, thus increasing the contribution total_amount - but what financial type should be used & how does it work when there are line items.

Note that if you are not actively vested in this at the moment I would consider closing the PR & tracking it through JIRA until you are.

**Comment by monishdeb** [comment]
Will reopen a new PR soon

**Pull request closed without merging**